### PR TITLE
Benchmark bvh construction as well as knn and radius searches

### PR DESCRIPTION
**Pull request by dalg24**
Partially addressing #429

### CODE REVIEW

File: docker/Dockerfile_stack [hunk 1]
```diff
@@ -234,6 +234,16 @@ RUN ["/bin/bash","-c","export KOKKOS_TOOLS_HASH=543c144bc64feb8b8cc4d8e05d04c0a0
...
 
+# Benchmark support library
+ENV BENCHMARK_DIR=/opt/benchmark
+RUN cd ${PREFIX} && \
+    git clone https://github.com/google/benchmark.git -b v1.4.1 && \
+    cd benchmark && \
+    git clone https://github.com/google/googletest.git -b release-1.8.1 && \
+    mkdir build && cd build && \
+    cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=${BENCHMARK_DIR} .. && \
+    make -j${NPROC} && make install && \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: @Rombur Note that the `compare.py` tool is not copied to the installation prefix.  We will have to handle it ourselves if we decide to use it.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 2]
```diff
@@ -99,145 +54,229 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-    std::chrono::duration<double> elapsed_seconds = end - start;
-    os << "construction " << elapsed_seconds.count() << "\n";
+template <class DeviceType>
+void BM_construction( benchmark::State &state )
+{
+    int const n_values = state.range( 0 );
+    auto bounding_boxes = constructBoxes<DeviceType>( n_values );
 
-    if ( perform_knn_search )
+    for ( auto _ : state )
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: This is straight out of google benchmark example/tutorial.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 3]
```diff
@@ -99,145 +55,230 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+#define REGISTER_BENCHMARK( DeviceType )                                       \
+    BENCHMARK_TEMPLATE( BM_construction, DeviceType )                          \
+        ->Arg( n_values )                                                      \
+        ->UseManualTime()                                                      \
+        ->Unit( benchmark::kMicrosecond );                                     \
+    BENCHMARK_TEMPLATE( BM_knn_search, DeviceType )                            \
+        ->Args( {n_values, n_queries, n_neighbors} )                           \
+        ->UseManualTime()                                                      \
+        ->Unit( benchmark::kMicrosecond );                                     \
+    BENCHMARK_TEMPLATE( BM_radius_search, DeviceType )                         \
```
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: So this lets you parameter sweep through the benchmark?

**Comment by aprokop** [comment]
I would prefer if there remained a way to run the example *without* the benchmark installed as it just adds an extra step for users.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 3 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: You can specify range for arguments.  See google/benchmark doc.

**Comment by dalg24** [comment]
> I would prefer if there remained a way to run the example without the benchmark installed as it just adds an extra step for users.

We can revisit that when we write up some sort of tutorial later.  I did not like the drivers as they were before and I do not want to have to maintain two versions at the moment.  These changes make this example much more suitable for measuring performance which was the only thing we used this driver for so far.

**Comment by dalg24** [comment]
> tag the new 18.09.0 image as latest and push it to Dockerhub, then drop 267008a before we merge

Done.  Will merge whenever CI builds are completed.

File: docker/README.md [hunk 4]
```diff
@@ -38,3 +38,7 @@ Clear things up about install prefix of LLVM OpenMP
 - Upgrade all LLVM packages to 6.0.0
 COMMENT: downloaded binaries for Clang + LLVM (including support for OpenMP) instead of using system package manager/building from source
 - Upgrade Open MPI to 2.1.3
+
+## 18.09.0 (September 5th, 2018)
+### TPLs addition
+- Add benchmark support library from google (version 1.4.1)
```
Comment (dalg24) [inline docker/README.md]: Actually realized I had forgotten to update the change log for our Docker container.

File: docker/Dockerfile_stack (hunk 1 above)
Comment (dalg24) [reply docker/Dockerfile_stack]: @sslattery Look at these instructions to install the benchmark support library.  And when you configure DTK/Trilinos, pass `-D CMAKE_PREFIX_PATH=<benchmark install prefix>`

**Pull request merged**